src/tls12.rs

```diff
@@ -0,0 +1,356 @@
+        match chacha20_poly1305_decrypt_in_place(
+            &self.key,
+            &nonce.0,
+            &auth_data,
+            &mut payload[..message_len],
+            &tag,
+        ) {
+            Ok(_) => {
+                payload.truncate(message_len);
+                Ok(msg.into_plain_message())
+            }
+            Err(symcrypt_error) => {
+                let custom_error_message = format!(
+                    "SymCryptError: {}",
+                    symcrypt_error.to_string() // Using general error to propagate the SymCrypt error back to the caller
+                );
+                return Err(rustls::Error::General(custom_error_message));
+            }
+        }
+    }
+}
+
+/// GCM 1.2
+/// Tls12Gcm impls [`Tls12AeadAlgorithm`].
+///
+/// [`algo_type`] represents either GCM128 or GCM256 which corresponds to a 16 and 32 byte key respectively.
+pub struct Tls12Gcm {
+    algo_type: AesGcm,
+}
+
+/// Gcm12Decrypt impls [`MessageDecrypter`]
+/// [`key`] is a [`GcmExpandedKey`] which takes in a key, and block type to return a Pin<Box<>>'d expanded key.
+/// The only supported block type is AES.
+/// [`iv`] is an implicit Iv that must be 4 bytes.
+pub struct Gcm12Decrypt {
+    key: GcmExpandedKey,
+    iv: [u8; 4],
+}
+
+/// Gcm12Encrypt impls [`MessageEncrypter`]
+/// [`key`] is a [`GcmExpandedKey`] which takes in a key, and block type to return a Pin<Box<>>'d expanded key.
+/// The only supported block type is AES.
+/// [`full_iv`] is a full_iv that includes both the implicit and the explicit iv.
+pub struct Gcm12Encrypt {
+    key: GcmExpandedKey,
+    full_iv: [u8; GCM_FULL_NONCE_LENGTH],
+}
+
+impl Tls12AeadAlgorithm for Tls12Gcm {
+    fn encrypter(&self, key: AeadKey, iv: &[u8], extra: &[u8]) -> Box<dyn MessageEncrypter> {
+        assert_eq!(iv.len(), 4);
+        assert_eq!(extra.len(), 8);
+        let mut full_iv = [0u8; GCM_FULL_NONCE_LENGTH];
+        full_iv[..4].copy_from_slice(iv);
+        full_iv[4..].copy_from_slice(extra);
+
+        // Unwrapping here, in the scenarios that GcmExpandKey would fail should result in a panic, ie: Not enough memory.
+        Box::new(Gcm12Encrypt {
+            key: GcmExpandedKey::new(key.as_ref(), BlockCipherType::AesBlock).unwrap(),
+            full_iv: full_iv,
+        })
+    }
+
+    fn decrypter(&self, key: AeadKey, iv: &[u8]) -> Box<dyn MessageDecrypter> {
+        assert_eq!(iv.len(), 4);
+        let mut implicit_iv = [0u8; 4];
+        implicit_iv.copy_from_slice(iv);
+
+        // Unwrapping here, in the scenarios that GcmExpandKey would fail should result in a panic, ie: Not enough memory.
+        Box::new(Gcm12Decrypt {
+            key: GcmExpandedKey::new(key.as_ref(), BlockCipherType::AesBlock).unwrap(),
+            iv: implicit_iv,
+        })
+    }
+
+    fn key_block_shape(&self) -> KeyBlockShape {
+        KeyBlockShape {
+            enc_key_len: self.algo_type.key_size(), // Can be either 16 or 32
+            fixed_iv_len: 4,
```

Comment:
>4 [](http://example.com/codeflow?start=26&length=1)

Seems like this 4 should also use a constant name?

Reply:
changed